packages/metadata/src/decorator-factory.ts

```diff
@@ -0,0 +1,539 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: @loopback/metadata
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Reflector} from './reflect';
+import * as _ from 'lodash';
+import * as debugModule from 'debug';
+const debug = debugModule('loopback:metadata:decorator');
+
+// tslint:disable:no-any
+
+function cloneDeep<T>(val: T): T {
+  if (val === undefined) {
+    return {} as T;
+  }
+  return _.cloneDeepWith(val, v => {
+    // Do not clone functions
+    if (typeof v === 'function') return v;
+    return undefined;
+  });
+}
+
+/**
+ * An object mapping keys to corresponding metadata
+ */
+export interface MetadataMap<T> {
+  [propertyOrMethodName: string]: T;
+}
+
+/**
+ * Options for a decorator
+ */
+export interface DecoratorOptions {
+  /**
+   * Inheritance will be honored
+   */
+  allowsInheritance?: boolean;
+  [name: string]: any;
+}
+
+/**
+ * Decorator function types
+ */
+export type DecoratorType =
+  | ClassDecorator
+  | PropertyDecorator
+  | MethodDecorator
+  | ParameterDecorator;
+
+/**
+ * Base factory class for decorator functions
+ *
+ * @example
+ * ```
+ * function classDecorator(spec: MySpec): ClassDecorator {
+ *   return ClassDecoratorFactory.createDecorator('my-key', spec);
+ * }
+ * ```
+ * or
+ * ```
+ * function classDecorator(spec: MySpec): ClassDecorator {
+ *   const factory: ClassDecoratorFactory<MySpec>('my-key', spec);
+ *   return factory.create();
+ * }
+ * ```
+ * These functions above declare `@classDecorator` that can be used as follows:
+ * ```
+ * @classDecorator({x: 1})
+ * class MyController {}
+ * ```
+ */
+export class DecoratorFactory<
+  T, // Type of the metadata spec for individual class/method/property/parameter
+  M extends T | MetadataMap<T> | MetadataMap<T[]>, // Type of the metadata
+  D extends DecoratorType // Type of the decorator
+> {
+  /**
+   * A symbol to reference the target of a decoration
+   */
+  static TARGET = Symbol('decorationTarget');
```

Comment:
I think we should not be using Symbols and use strings instead. Here is the use case where Symbols don't work:

 1. There are two modules `EXT` and `APP`, each depending on a different version of  `@loopback/metadata`. As a result, there are two copies of `@loopback/metadata` in `node_modules` tree. As a result, there are two `TARGET` symbols inside the Node.js process, one symbol is used by EXT and another symbol is used by APP.

 2. APP module sets metadata using its `TARGET` symbol.

 3. EXT module wants to read metadata set by APP. Because it has different `TARGET` symbol, it cannot read information set by APP.